cpp/ConnectionPool.cpp

```diff
@@ -0,0 +1,330 @@
+                                        string const &query) {
+  sqlite3 *db = nullptr;
+  if (writeConnection.currentLockId == contextId) {
+    db = writeConnection.connection;
+  } else {
+    // Check if it's a read connection
+    for (int i = 0; i < maxReads; i++) {
+      if (readConnections[i].currentLockId == contextId) {
+        db = readConnections[i].connection;
+        break;
+      }
+    }
+  }
+  if (db == nullptr) {
+    // throw error that context is not available
+    return SequelLiteralUpdateResult{
+        .type = SQLiteError,
+        .message = "Context is no longer available",
+    };
+  }
+
+  return sqliteExecuteLiteralWithDB(db, query);
+}
+
+void ConnectionPool::setOnContextAvailable(void (*callback)(std::string,
+                                                            ConnectionLockId)) {
+  onContextCallback = callback;
+}
+
+void ConnectionPool::setTableUpdateHandler(
+    void (*callback)(void *, int, const char *, const char *, sqlite3_int64)) {
+  // Only the write connection can make changes
+  sqlite3_update_hook(writeConnection.connection, callback,
+                      (void *)(dbName.c_str()));
+}
+
+void ConnectionPool::closeContext(ConnectionLockId contextId) {
+  if (writeConnection.currentLockId == contextId) {
+    if (writeQueue.size() > 0) {
+      // There are items in the queue, activate the next one
+      activateContext(&writeConnection, writeQueue[0]);
+      writeQueue.erase(writeQueue.begin());
+    } else {
+      // No items in the queue, clear the context
+      writeConnection.currentLockId = EMPTY_LOCK_ID;
+    }
+  } else {
+    // Check if it's a read connection
+    for (int i = 0; i < maxReads; i++) {
+      if (readConnections[i].currentLockId == contextId) {
+        if (readQueue.size() > 0) {
+          // There are items in the queue, activate the next one
+          activateContext(&readConnections[i], readQueue[0]);
+          readQueue.erase(readQueue.begin());
+        } else {
+          // No items in the queue, clear the context
+          readConnections[i].currentLockId = EMPTY_LOCK_ID;
+        }
+        return;
+      }
+    }
+  }
+}
+
+void ConnectionPool::closeAll() {
+  sqlite3_close_v2(writeConnection.connection);
+  for (int i = 0; i < maxReads; i++) {
+    sqlite3_close_v2(readConnections[i].connection);
+  }
+}
+
+SQLiteOPResult ConnectionPool::attachDatabase(std::string const dbFileName,
+                                              std::string const docPath,
+                                              std::string const alias) {
+
+  /**
+   * There is no need to check if mainDBName is opened because
+   * sqliteExecuteLiteral will do that.
+   * */
+  string dbPath = get_db_path(dbFileName, docPath);
+  string statement = "ATTACH DATABASE '" + dbPath + "' AS " + alias;
+
+  auto dbConnections = getAllConnections();
+
+  for (auto &connection : dbConnections) {
+    SequelLiteralUpdateResult result =
```

Comment:
This bypasses the locking logic - could result in concurrency issues.

Reply:
This currently operates on all the connections. Adding lock protections to await for when all the connections are available would be tricky. This would most likely need to be executed before any locks are requested. We could add a check to ensure that there are no active locks on any of the connections and fail if there are. Do you think that would be sufficient?

Reply:
With the attach call being synchronous, that should be sufficient. I'd assume that in most cases this would be called before any other transactions.

If it's changed to an asynchronous call, we'd need to revisit.

Reply:
Added a check to ensure all connections are free.